### DISCUSSION TITLE
Pairing monaco-editor with JDTLS using monaco-languageclient

### DISCUSSION THREAD

**Discussion by zeekliviu** [TypeFox/monaco-languageclient#700] [Q&A] [answered]
Hello, and thanks for stopping by here!

I have been trying to pair those two components for days and failed. My use case is that I need to render the monaco-editor in a React component for solving Java problems for the e-learning platform I am trying to build for my bachelor thesis. I need to pair monaco with JDTLS to have the features that VS Code has (i.e. informative tooltips, autocomplete, autoimport, etc.) because those are helpful for beginners.

I have already built a server for the RESTful API and tried to bundle it with the ws as seen in the example.

I have followed the example on Groovy from [here](https://github.com/TypeFox/monaco-languageclient/tree/main/packages/examples/src/groovy) and made a [minimal reproducible project](https://github.com/zeekliviu/react-monaco-with-jdtls). The command to start both the front and back end is `npm run dev`.

The error I am facing is that the editor seems to not be configured correctly and I honestly do not know what to do next.

What am I missing?

Thanks for any help! 🙏🏻

**Comment by kaisalmen** [comment]
Hi @zeekliviu thank you for the reproducible example. I will look into this likely tomorrow and provide feedback.

**Reply by kaisalmen** [reply, marked as answer]
@zeekliviu I have created a working version on this branch: https://github.com/TypeFox/monaco-languageclient/tree/jdtls

More details later, especially what were the obstacles to make it work properly

**Reply by zeekliviu** [reply]
Thank you, sir! 🙏🏻 

I will patiently wait for you to describe the adventure of getting it working properly. I bet it's a one of a kind story!

**Reply by kaisalmen** [reply]
@zeekliviu see PR #708

**Reply by kaisalmen** [reply]
@zeekliviu from the PR description "JDT LS requires that the files physically exist at the specified location. An openTextDocument is not sufficient. This works flawlessly with other language servers"
This was the main obstacle. The container ensure the server starts properly on all platforms with everything a the correct location.

**Reply by zeekliviu** [reply]
@kaisalmen all right; so from my use case, there will be a need to modify the starting code file inside the container before a client connects to always reflect the boilerplate code for each problem. I can not really start a container for each individual as that would kill my low spec server. Maybe there is a solution with different workspaces? I'll look into it.

I want to thank you and to all of the team for the effort you put into this project!

If there is something I can do in return, let me know!

**Reply by kaisalmen** [reply]
> I can not really start a container for each individual as that would kill my low spec server. Maybe there is a solution with different workspaces? I'll look into it.

Yes, different workspaces could help, but the general problem is in a distributed case, that server and client don't share the same file system and that each language server implementation can behave differently with regarding to opening files and storing/handling files. A mechanism would be needed to transfer sets of files to the server (initially).

**Reply by zeekliviu** [reply]
Hi, @kaisalmen 

I have almost managed to getting it working as a React component. Here is the [reproducible example](https://github.com/zeekliviu/react-with-jdtls). With the JDTLS container started on 30003, just run `npm i && npm run dev`.

I got one little tiny question: why the theme and highlighting is not working as expected?

![image](https://github.com/user-attachments/assets/7752cfd7-187f-4955-a1ca-bb4d7e2af8c1)
![image](https://github.com/user-attachments/assets/d31bad8a-f090-45bf-a915-01b11ad9b188)

EDIT: Oh, I see that this problem is only in development. After the build the syntax highlighting is present and the theme is applied. Thanks for all!

Many thanks!

**Reply by kaisalmen** [reply]
@zeekliviu the problem during development should disappear if you use this plugin: https://github.com/TypeFox/monaco-languageclient/blob/main/vite.config.ts#L58 You have to add the dependency as devDependency.

**Reply by ivan-kzn** [reply]
Hi! @kaisalmen thanks a lot for the working example! one question regarding jdtls. As I can see from code, each client, which requires connection to ls creates it's own process of jdtls in docker container, it requires huge amount of resources, when you have e.g. 10 clients or more... is there any chance to reduce resource usage for multiple clients? Maybe one instance can be reused by clients connected to same document

**Reply by kaisalmen** [reply]
Hi @ivan-kzn usually language servers run in 1to1 relation with a client locally. There are projects like https://github.com/CodinGame/languageserver-mutualized that help to overcome this., but there is always a trade-off. We offer example that show how to get started, but don't have the resources to do more. Also, what is potentially possible varies of course by each language server implementation.
If you have a server that runs in the browser via JS or WASM you get rid off the problem again, because everything is executed on the client side. But like for jdtls this is not possible right now.

**Reply by ivan-kzn** [reply]
hi @kaisalmen thanks for your reply!
I understand the reason why it is running 1to1 locally, but when talking about web development and lot's of clients trying to connect to same LS it sounds useless...
It doesn't minimize your great efforts in monaco languageclient and working example for java ls, just thinking how this can be overcome. Thanks for the link, I'll take a look at codingame foundings

**Reply by kaisalmen** [reply]
@ivan-kzn yes, scaling this is not an easy task and goes far beyond what we can offer here, but it helps to see the start of the road that climbs up the mountain. 🙂

**Reply by CGNonofr** [reply]
I love poetry (:
